File: locus-api-android/src/main/java/locus/api/android/ActionDisplayTracks.java
package locus.api.android;

import java.util.List;

import locus.api.android.utils.LocusConst;
import locus.api.android.utils.exceptions.RequiredVersionMissingException;
import locus.api.objects.Storable;
import locus.api.objects.extra.Track;
import locus.api.utils.Logger;
import android.content.Context;
import android.content.Intent;

public class ActionDisplayTracks extends ActionDisplay {

	private static final String TAG = ActionDisplayTracks.class.getSimpleName();
	
	// SEND ONE SINGLE TRACK
	
	public static boolean sendTrack(Context context,
			Track track, ExtraAction extraAction)
					throws RequiredVersionMissingException {
		return sendTrack(LocusConst.ACTION_DISPLAY_DATA,
				context, track, extraAction == ExtraAction.IMPORT,
						extraAction == ExtraAction.CENTER, false);
	}
	
	public static boolean sendTrack(Context context, 
			Track track, ExtraAction extraAction, boolean startNavigation)
					throws RequiredVersionMissingException {
		return sendTrack(LocusConst.ACTION_DISPLAY_DATA, 
				context, track, extraAction == ExtraAction.IMPORT,
						extraAction == ExtraAction.CENTER, startNavigation);
	}
	
	public static boolean sendTrackSilent(Context context, 
			Track track, boolean centerOnData)
					throws RequiredVersionMissingException {
		return sendTrack(LocusConst.ACTION_DISPLAY_DATA_SILENTLY, 
				context, track, false, centerOnData, false);
	}
	
	private static boolean sendTrack(String action, Context context, Track track,
			boolean callImport, boolean centerOnData, boolean startNavigation)
			throws RequiredVersionMissingException {
		// check track
		if (track == null || track.getPoints().size() == 0) {
			Logger.logE(TAG, "sendTrack(" + action + ", " + context + ", " + track + ", " +
					callImport + ", " + centerOnData + ", " + startNavigation + "), " +
					"track is null or contain no points");
			return false;
		}
		
		// create and start intent
		Intent intent = new Intent();
		intent.putExtra(LocusConst.INTENT_EXTRA_TRACKS_SINGLE, track.getAsBytes());
		intent.putExtra(LocusConst.INTENT_EXTRA_START_NAVIGATION, startNavigation);
		return sendData(action, context, intent, callImport, centerOnData);
	}
	
	// SEND TRACK PACK (MORE THEN ONE)
	
	public static boolean sendTracks(Context context,
			List<Track> tracks, ExtraAction extraAction)
					throws RequiredVersionMissingException {
		return sendTracks(LocusConst.ACTION_DISPLAY_DATA, 
				context, tracks, extraAction == ExtraAction.IMPORT,
				extraAction == ExtraAction.CENTER);
	}
	
	public static boolean sendTracksSilent(Context context,
			List<Track> tracks, boolean centerOnData)
					throws RequiredVersionMissingException {
		return sendTracks(LocusConst.ACTION_DISPLAY_DATA_SILENTLY,
				context, tracks, false, centerOnData);
	}
	
	private static boolean sendTracks(String action, Context context,
			List<Track> tracks, boolean callImport, boolean centerOnData)
			throws RequiredVersionMissingException {
		// check data
		if (tracks == null || tracks.size() == 0) {
			return false;
		}
		
		// create and start intent
		Intent intent = new Intent();
		intent.putExtra(LocusConst.INTENT_EXTRA_TRACKS_MULTI, 
				Storable.getAsBytes(tracks));
		return sendData(action, context, intent, callImport, centerOnData);
	}
}

File: locus-api-android/src/main/java/locus/api/android/ActionDisplayTracks.kt
package locus.api.android

import android.content.Context
import locus.api.android.features.sendToApp.SendMode
import locus.api.android.features.sendToApp.tracks.SendTrack
import locus.api.android.features.sendToApp.tracks.SendTracks
import locus.api.android.utils.LocusConst
import locus.api.android.utils.exceptions.RequiredVersionMissingException
import locus.api.objects.geoData.Track

@Deprecated(message = "Use `SendTrack` or `SendTracks` objects.")
object ActionDisplayTracks {

    // tag for logger
    private const val TAG = "ActionDisplayTracks"

    // SEND ONE SINGLE TRACK

    @Throws(RequiredVersionMissingException::class)
    fun sendTrack(ctx: Context, track: Track,
            extraAction: ActionDisplayVarious.ExtraAction,
            startNavigation: Boolean = false): Boolean {
        return sendTrack(LocusConst.ACTION_DISPLAY_DATA,
                ctx, track, extraAction == ActionDisplayVarious.ExtraAction.IMPORT,
                extraAction == ActionDisplayVarious.ExtraAction.CENTER, startNavigation)
    }

    @Throws(RequiredVersionMissingException::class)
    fun sendTrackSilent(ctx: Context, track: Track, centerOnData: Boolean): Boolean {
        return sendTrack(LocusConst.ACTION_DISPLAY_DATA_SILENTLY,
                ctx, track, false, centerOnData, false)
    }

    @Throws(RequiredVersionMissingException::class)
    private fun sendTrack(action: String, ctx: Context, track: Track,
            callImport: Boolean, centerOnData: Boolean, startNavigation: Boolean): Boolean {
        // prepare mode
        val sendMode = when {
            callImport -> {
                SendMode.Import()
            }
            action == LocusConst.ACTION_DISPLAY_DATA -> {
                SendMode.Basic(centerOnData)
            }
            action == LocusConst.ACTION_DISPLAY_DATA_SILENTLY -> {
                SendMode.Silent()
            }
            else -> {
                return false
            }
        }

        // send request
        return SendTrack(sendMode, track) {
            this.startNavigation = startNavigation
        }.send(ctx)
    }

    // SEND TRACK PACK (MORE THEN ONE)

    @Throws(RequiredVersionMissingException::class)
    fun sendTracks(ctx: Context, tracks: List<Track>,
            extraAction: ActionDisplayVarious.ExtraAction): Boolean {
        return sendTracks(LocusConst.ACTION_DISPLAY_DATA,
                ctx, tracks, extraAction == ActionDisplayVarious.ExtraAction.IMPORT,
                extraAction == ActionDisplayVarious.ExtraAction.CENTER)
    }

    @Throws(RequiredVersionMissingException::class)
    fun sendTracksSilent(ctx: Context, tracks: List<Track>, centerOnData: Boolean): Boolean {
        return sendTracks(LocusConst.ACTION_DISPLAY_DATA_SILENTLY,
                ctx, tracks, false, centerOnData)
    }

    @Throws(RequiredVersionMissingException::class)
    private fun sendTracks(action: String, ctx: Context,
            tracks: List<Track>, callImport: Boolean, centerOnData: Boolean): Boolean {
        // prepare mode
        val sendMode = when {
            callImport -> {
                SendMode.Import()
            }
            action == LocusConst.ACTION_DISPLAY_DATA -> {
                SendMode.Basic(centerOnData)
            }
            action == LocusConst.ACTION_DISPLAY_DATA_SILENTLY -> {
                SendMode.Silent()
            }
            else -> {
                return false
            }
        }

        // send request
        return SendTracks(sendMode, tracks)
                .send(ctx)
    }
}
